src/shared/serverinfo.cpp

```diff
@@ -0,0 +1,507 @@
+#include <algorithm>
+#include <assert.h>
+#include <limits.h>
+#include <stdio.h>
+
+#include <zlib.h>
+#include <SDL.h>
+#include <enet/enet.h>
+
+#ifndef STANDALONE
+  #ifdef __APPLE__
+    #define GL_GLEXT_LEGACY
+    #define __glext_h_
+    #include <OpenGL/gl.h>
+    #define main SDL_main
+  #else
+    #include <SDL_opengl.h>
+  #endif
+#endif
+
+using std::swap;
+#include "tools.h"
+#include "command.h"
+#include "serverinfo.hpp"
+
+void serverinfo::clearpings()
+{
+    m_ping = WAITING;
+    loopk(MAXPINGS)
+    {
+        m_pings[k] = WAITING;
+    }
+    m_nextping = 0;
+    m_lastping = m_lastinfo = -1;
+}
+
+void serverinfo::cleanup()
+{
+    clearpings();
+    m_attr.clear();
+    m_players.clear();
+    m_handles.clear();
+}
+
+void serverinfo::addping(int rtt, int millis)
+{
+    if(millis >= m_lastping)
+    {
+        m_lastping = -1;
+    }
+    m_pings[m_nextping] = rtt;
+    m_nextping = (m_nextping+1)%MAXPINGS;
+    int numpings = 0;
+    int totalpings = 0;
+    loopk(MAXPINGS)
+    {
+        if(m_pings[k] != WAITING)
+        {
+            totalpings += m_pings[k];
+            numpings++;
+        }
+    }
+    m_ping = numpings ? totalpings/numpings : WAITING;
+}
+
+// return description if not empty, name otherwise.
+const char* serverinfo::description( void ) const
+{
+    return m_sdesc[0] ? m_sdesc : m_name;
+}
+
+// returns protocol version
+//
+// -1 if server still have no informations
+// 0x7FFF if server uses VERSION_GAME on m_attr[0] (for older protocol compat?)
+// m_attr[0] clamped in range [0,0x7FFE]
+int serverinfo::protocol_version( void ) const
+{
+    if( false
+        || m_address.host == ENET_HOST_ANY
+        || m_ping >= serverinfo::WAITING
+        || m_attr.empty()
+      )
+    {
+        return -1;
+    }
+    return m_attr[0] == VERSION_GAME ? 0x7fff : clamp( m_attr[0], 0, 0x7FFF - 1 );
+}
+
+// return "status" of server, aka: can it be joined
+SSTAT serverinfo::server_status( void ) const
+{
+    if( m_attr.size() > 4 && m_players.size() >= m_attr[4] )
+    {
+        return SSTAT_FULL;
+    }
+    if(m_attr.size() <= 5)
+    {
+        return SSTAT_UNKNOWN;
+    }
+    switch(m_attr[5])
+    {
+        case MM_LOCKED:
+            return SSTAT_LOCKED;
+        case MM_PRIVATE:
+        case MM_PASSWORD:
+            return SSTAT_PRIVATE;
+        default:
+            return SSTAT_OPEN;
+    }
+}
+
+// start of public methods definitions
+
+serverinfo::serverinfo(uint ip, int port, int priority)
+ : m_port(port), m_priority(priority), m_resolved(ip==ENET_HOST_ANY ? UNRESOLVED : RESOLVED)
+{
+    m_name[0] = m_map[0] = m_sdesc[0] = m_authhandle[0] = m_flags[0] = m_branch[0] = '\0';
+    m_address.host = ip;
+    m_address.port = port+1;
+    clearpings();
+}
+
+serverinfo::~serverinfo()
+{
+    cleanup();
+}
+
+void serverinfo::reset()
+{
+    m_lastping = m_lastinfo = -1;
+    s_sortedservers = false;
+}
+
+int serverinfo::compare( serverinfo const& other, int style, bool reverse ) const
```

Comment:
We usually put `const` on the left side when possible. It's confusing to mix the two styles. This is done in some other places too, not only this line.

Reply:
The reason is a different one. This use of `const` indicates that the method does not alter the state of the object. That means that the compiler permits its execution also for `const` references. You need to do the same for comparison operators as well. There is no way to put it on the left side.

Reply:
I was talking about the `const` inside the argument list, not after the function. We would normally do `const serverinfo &other`, not `serverinfo const& other`

Reply:
I see. Well, in C++ it's nowadays more or less a matter of personal preference. In good old C times, the order did make a huge difference. Simple example:

```cpp
// one can modify the contents of a
char * const a;
// here, one cannot change the contents, but change the pointer address
const char* b;
```

I'm personally in preference of `const serverinfo& abc`. But we have an issue about code style open anyway. We've agreed on avoiiding such "preference style" discussions until we have a consensus (and ideally a `clang-format` file which can be used to auto-format files to some extent).

Reply:
Yes, order matters if you put two things on the left side of the `const` but not if you put `const` on the left or right side of the leftmost thing.